Rework Prometheus metrics API - counters vs gauges

cubranic opened the pull request:
As discussed in #353

andrerenard commented on lib/core/prometheusMetrics.hpp:
```diff
@@ -16,6 +16,116 @@ using std::string;
 
 namespace kotekan {
```
I wonder if this should have it's own namespace.  Just because `kotekan::Family` and `kotekan::Metric` seems overly generic.   I would either subname space this, or add some indication this is for Prometheus to all the names.

cubranic replied:
I'm inclined to go with the subname: either `kotekan::prometheus::` or `kotekan::metrics`. It would probably make sense to move the `prometheusMetrics` class into it as well, no?

andrerenard replied:
I would go with `kotekan::prometheus::`, the word "metrics" might well apply to other systems of tracking stats.

cubranic replied:
Sounds good. And now `kotekan::prometheusMetrics` becomes `kotekan::prometheus::Metrics`.

andrerenard commented on lib/core/prometheusMetrics.hpp:
```diff
@@ -16,6 +16,116 @@ using std::string;
+class Counter : public Metric {
+public:
+    Counter(const std::vector<string>&);
+    void inc();
```
Tempted to operator overload `++` here.

cubranic replied:
I'm leaving it alone for now, since `.inc()` is used in other Prometheus clients. Let's discuss this and other ways to improve the API from developers' point after this cycle.

cubranic commented:
I consciously modelled the new API on other Prometheus clients, where declaring and using the metric is done in separate spots. This also catches duplicate declarations, and throws an error. But we could have a shim that follows the current API, where there is single function that both looks-up/declares a metric and set its current value. 

Davor

> On Jun 27, 2019, at 8:43 AM, Andre Renard <notifications@github.com> wrote:
> 
> @andrerenard requested changes on this pull request.
> 
> Some comments inline. I'm a bit concerned by adding extra work (lines of code) in setting up metrics now, especially metric families, but I also don't see a clear way to reduce it at the moment. It might just be we need the extra code to express this.
> 
> In lib/core/prometheusMetrics.hpp:
> 
> > @@ -16,6 +16,116 @@ using std::string;
>  
>  namespace kotekan {
>  
> +/**
> + * @class metric
> + * @brief An internal base class for storing metric values
> + */
> +class metric {
> There seem to be lowercase and Uppercase versions of objects here. I think you only mean to have one.
> 
> In lib/dpdk/iceBoardShuffle.hpp:
> 
> > +                                        unique_name, {"port"})),
> +    second_shuffle_error(kotekan::prometheusMetrics::instance()
> +                         .AddGauge("kotekan_dpdk_shuffle_fpga_second_stage_shuffle_errors_total",
> +                                   unique_name, {"port", "fpga_lane"})),
> +    second_crc_error(kotekan::prometheusMetrics::instance()
> +                     .AddGauge("kotekan_dpdk_shuffle_fpga_second_stage_crc_errors_total",
> +                               unique_name, {"port"})),
> +    second_missing_short_error(kotekan::prometheusMetrics::instance()
> +                               .AddGauge("kotekan_dpdk_shuffle_fpga_second_stage_missing_short_errors_total",
> +                                         unique_name, {"port"})),
> +    second_long_error(kotekan::prometheusMetrics::instance()
> +                      .AddGauge("kotekan_dpdk_shuffle_fpga_second_stage_long_errors_total",
> +                                unique_name), {"port"}),
> +    second_fifo_overflow_error(kotekan::prometheusMetrics::instance()
> +                               .AddGauge("kotekan_dpdk_shuffle_fpga_second_stage_fifo_overflow_errors_total",
> +                                         unique_name, {"port"})) {
> ⬇️ Suggested change
> -                                         unique_name, {"port"})) {
> +                                         unique_name, {"port", "fpga_lane"})) {
> In lib/core/prometheusMetrics.hpp:
> 
> > @@ -16,6 +16,116 @@ using std::string;
>  
>  namespace kotekan {
> I wonder if this should have it's own namespace. Just because kotekan::Family and kotekan::Metric seems overly generic. I would either subname space this, or add some indication this is for Prometheus to all the names.
> 
> In lib/core/prometheusMetrics.hpp:
> 
> > + */
> +class Metric {
> +public:
> +    Metric(const std::vector<string>&);
> +    virtual ~Metric() = default;
> +
> +    /// @brief Returns the stored value as a string.
> +    virtual string to_string() = 0;
> +    virtual std::ostringstream& to_string(std::ostringstream& out) = 0;
> +    const std::vector<string> label_values;
> +};
> +
> +class Counter : public Metric {
> +public:
> +    Counter(const std::vector<string>&);
> +    void inc();
> Tempted to operator overload ++ here.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

andrerenard commented:
I see the point about catching duplicate declarations, we have had that happen in the past.  So let's not bother with the extra function to lookup/declare for now.   However if possible it would be good to address the other comments before we merge, mostly around naming.  And if possible a fix to avoid the need for `std::to_string()` would be nice, but not critical for this release.

cubranic commented:
I'll work on this today -- it should be pretty mechanical and quick to do.

cubranic commented on lib/core/prometheusMetrics.hpp:
```diff
@@ -16,6 +16,116 @@ using std::string;
+ * @class metric
+ * @brief An internal base class for storing metric values
+ */
+class metric {
```
The lower case ones were used in the old API. I removed them now.

cubranic commented on lib/dpdk/iceBoardShuffle.hpp:
```diff
@@ -169,7 +169,37 @@ class iceBoardShuffle : public iceBoardHandler {
+                                unique_name), {"port"}),
+    second_fifo_overflow_error(kotekan::prometheusMetrics::instance()
+                               .AddGauge("kotekan_dpdk_shuffle_fpga_second_stage_fifo_overflow_errors_total",
+                                         unique_name, {"port"})) {
```
Can you clarify the suggested change here? `kotekan_dpdk_shuffle_fpga_second_stage_fifo_overflow_errors_total` only uses the "port" label. Only `...second_stage_shuffle_errors_total` and `...third_stage_shuffle_errors_total` also add the `fpga_lane` label.

cubranic commented:
Well, it wasn't that mechanical, but I've finalized the API renaming, and hopefully addressed all of your points.

One thing I just realized is that we should probably change the type of the "value" of the counter from `int` to `uint64_t` since they can't be negative (start at 0 and count up), or maybe `double`, since that's the underlying storage type in Prometheus.

andrerenard commented:
Just need to fix the formatting issue (see `#kotekan_dev` on slack), but otherwise this looks good to me.
